pkg/ruler/api.go

```diff
@@ -551,3 +551,28 @@ func (a *API) DeleteRuleGroup(w http.ResponseWriter, req *http.Request) {
 
 	respondAccepted(w, logger)
 }
+
+func (a *API) ListAllRules(w http.ResponseWriter, req *http.Request) {
```

Comment:
Similar to listing alert groups, I think we should
1) get all users
2) get rule groups for batch of users
3) send this partial output, and then go to step 2 and repeat

Otherwise loading all rule groups into memory first may take a while, and its memory consumption is not bounded in any way. WDYT?

Reply:
Yes, I agree with you. 

But the ruler store have a ListAllRuleGroups interface which returns all rule groups for all users.

Reply:
Ruler sstore have no interface like GetAlertConfigs which allow get configs by input batch users.

Is there any plan to support such a interface? I can work for that, but it may lasts a long time.

Reply:
> But the ruler store have a ListAllRuleGroups interface which returns all rule groups for all users.

I don't think we should use it -- it suffers the same issue as described above -- loading all configs into memory could cause OOM. Ruler also doesn't do that when sharding is enabled.

I would not advocate for using `GetAlertConfigs` for *ALL* users at once. That is why I suggested to use "batches" of users instead.

Reply:
Get it.

Reply:
Also updated.

Now the api use iterator mode.

There is a problem:

```
// for k, v := range tt.value {
// 	iter.Put(map[string]*testStruct{k: v})
// }
```

TestStreamWriteYAMLResponse use iter not work, but TestAMConfigListUserConfig and TestRuler_ListAllRules work well, suspected that httptest flush is not a real flush.

Reply:
This problem have resolved and the reson is '---'.

Yaml unmarshal will stop when find '---'.

Reply:
Similar to listing alert groups. The `partial output` seems good now.